### PR TITLE
 Refactor resources "dockerimage" and "dockerdigest" into a new package "dockerimage"

### PR DESCRIPTION
**Pull request by dduportal**
This PR comes from the work in #412 .

It introduces the following changes:

## Functional changes:

- Use the Registry API v2 for all dockerdigest and dockerimage calls:
  * Deprecates the "Bearer Token" specification in favor of username + password (or anonymous), as it's generated automatically as part of the V2 API
  * Fully support Architecture for images (bugfix for DockerHub images, new feature for quay.io images)

## Technical changes:

- Add manifest examples to allow (manual) end to end testing for both providers
- Introduce a method `New(Spec)` to allow initializing an object with a state

- Introduces a utility `docker` package with sub-packages
  * Defines the default values for image naming
  * Introduces `docker/dockerimage` sub-package to manage Docker images parsing logic
  * Introduces `docker/dockerregistry` sub-package to manage registry API logic

- Introduces packages to allow unit testing with mocks from HTTP clients and Docker utilities:
  * Introduces `docker/dockermocks` sub-package to provides mocks for DockerRegistry logic
  * Introduces `pkg/core/httpclient` package to define HTTP client interface (and allows mocking it on each package)

- Refactor dockerdigest resource:
  * In its own package, using the docker/* utilities
  * Add a `New` constructor method
  * Add unit tests with >80% code coverage

- Refactor dockerimage resource:
  * In its own package, using the docker/* utilities
  * Add a `New` constructor method
  * Add unit tests with >80% code coverage

## Test

To test this pull request, you can run the following commands:

```shell
go build -o dist/updatecli
./dist/updatecli diff --config ./examples/updatecli.generic/dockerdigest.yaml
./dist/updatecli diff --config ./examples/updatecli.generic/dockerimage.yaml

```

### CODE REVIEW

**Comment by olblak** [comment]
While you are into this, you may have a look to https://github.com/updatecli/updatecli/issues/432#issuecomment-1000694620

**Comment by dduportal** [comment]
> While you are into this, you may have a look to [#432 (comment)](https://github.com/updatecli/updatecli/issues/432#issuecomment-1000694620)

Yep, good thinking! Answered: https://github.com/updatecli/updatecli/issues/432#issuecomment-1000703943

**Comment by dduportal** [comment]
Ping @olblak this one is ready for review \o/

**Comment by dduportal** [comment]
@olblak this PR does not have a specific issue, so I wasn't able to propose a milestone.

WDYT about 0.18.0 ?

**Comment by olblak** [comment]
> @olblak this PR does not have a specific issue, so I wasn't able to propose a milestone.
> 
> WDYT about 0.18.0 ?

Yep let exceptionally set the milestone on the pullrequest

File: examples/updatecli.generic/dockerdigest.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    name: "Get digest of the latest jenkins/jenkins:lts image on the DockerHub"
+    spec:
+      image: jenkins/jenkins
+      tag: lts
+      architecture: "amd64"
+      username: "dduportal"
+      password: "{{ requiredEnv `DOCKERHUB_TOKEN` }}"
+  nginxOfficialDockerHub:
+    kind: dockerDigest
+    name: "Get digest of the latest jenkins/jenkins:lts image on the DockerHub"
```
Comment (olblak) [inline examples/updatecli.generic/dockerdigest.yaml:22]:
```suggestion
    name: "Get digest of the latest nginx:alpine image on the DockerHub"
```

**Comment by olblak** [comment]
This pullrequest is a very nice refactoring especially considering the lack of documentation around docker library

**Review by olblak — APPROVED** [review]

**Comment by dduportal** [comment]
@olblak can you approve-it again? It's in "auto merge" but by adding your suggestion cancelled your previous approval

**Review by olblak — APPROVED** [review]
:shipit:

**Pull request merged**